wlauto/devices/android/gem5/__init__.py

```diff
@@ -0,0 +1,820 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+# Original implementation by Rene de Jong. Updated by Sascha Bischoff.
+
+import logging
+import os
+import re
+import shutil
+import socket
+import subprocess
+import sys
+import tarfile
+import time
+from pexpect import EOF, TIMEOUT
+
+from wlauto import AndroidDevice, settings, Parameter
+from wlauto.core import signal as sig
+from wlauto.utils import ssh
+from wlauto.exceptions import DeviceError, ConfigError
+
+
+class Gem5Device(AndroidDevice):
+    """
+    Implements gem5 Android device.
+
+    This class allows a user to connect WA to a simulation using gem5. The
+    connection to the device is made using the telnet connection of the
+    simulator, and is used for all commands. The simulator does not have ADB
+    support, and therefore we need to fall back to using standard shell
+    commands.
+
+    Files are copied into the simulation using a VirtIO 9P device in gem5. Files
+    are copied out of the simulated environment using the m5 writefile command
+    within the simulated system.
+
+    When starting the workload run, the simulator is automatically started by
+    Workload Automation, and a connection to the simulator is established. WA
+    will then wait for Android to boot on the simulated system (which can take
+    hours), prior to executing any other commands on the device. It is also
+    possible to resume from a checkpoint when starting the simulation. To do
+    this, please append the relevant checkpoint commands from the gem5
+    simulation script to the gem5_discription argument in the agenda.
+
+    Host system requirements:
+        * VirtIO support. We rely on diod on the host system. This can be
+          installed on ubuntu using the following command:
+
+                sudo apt-get install diod
+
+    Guest requirements:
+        * VirtIO support. We rely on VirtIO to move files into the simulation.
+          Please make sure that the following are set in the kernel
+          configuration:
+
+                CONFIG_NET_9P=y
+
+                CONFIG_NET_9P_VIRTIO=y
+
+                CONFIG_9P_FS=y
+
+                CONFIG_9P_FS_POSIX_ACL=y
+
+                CONFIG_9P_FS_SECURITY=y
+
+                CONFIG_VIRTIO_BLK=y
+
+        * m5 binary. Please make sure that the m5 binary is on the device and
+          can by found in the path.
+        * Busybox. Due to restrictions, we assume that busybox is installed in
+          the guest system, and can be found in the path.
+    """
+
+    name = 'gem5_android'
+    default_working_directory = '/data/'
+    has_gpu = False
+    platform = 'android'
+    path_module = 'posixpath'
+    default_package_data_directory = '/data/data/'
+    default_binaries_directory = '/system/bin'
+    default_external_storage_directory = '/data/data/'
+    default_adb_name = None
```

Comment:
The above `default_`'s should not be necessary as they will be handled by the super